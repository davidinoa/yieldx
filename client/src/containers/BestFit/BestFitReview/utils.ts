import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
