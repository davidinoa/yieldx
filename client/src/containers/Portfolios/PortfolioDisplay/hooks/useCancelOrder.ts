import {
  OrderStatus,
  usePatchPortfolioOrderCancelMutation
} from "providers/graphql/hooks";

function useCancelOrder() {
  const [cancelOrder] = usePatchPortfolioOrderCancelMutation();
  return (id: string, portfolioId: string) =>
    cancelOrder({
      variables: {
        id,
        portfolioId,
        cancelOrderRequestInput: {}
      },
      update(cache) {
        cache.modify({
          id: `Order2:${id}`,
          fields: {
            status() {
              return OrderStatus.Canceled;
            }
          }
        });
      }
    });
}

export default useCancelOrder;
