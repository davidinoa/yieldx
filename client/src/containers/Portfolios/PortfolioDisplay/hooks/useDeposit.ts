import {
  usePostInvestorPortfolioInternalPortfolioFundTransferMutation,
  Position2
} from "providers/graphql/hooks";

import YieldXPortfolio from "models/YieldXPortfolio";

interface Params {
  investorId: string;
  amount: number;
  cashPortfolio: {
    cashPos: Position2;
    cashAvailable: number;
    portfolioId: string;
  };
  portfolio: YieldXPortfolio;
}

function useDeposit() {
  const [
    makeTransfer
  ] = usePostInvestorPortfolioInternalPortfolioFundTransferMutation();

  return ({ investorId, amount, portfolio, cashPortfolio }: Params) => {
    const { cashAvailable } = cashPortfolio;
    const cashPosition = portfolio.positions.find(
      position => position.id === "CASH-USD" || position.id === "USD"
    );
    const cashAvailableInPortfolio =
      (cashPosition?.unsettled ?? 0) > 0
        ? cashPosition?.settled ?? 0
        : cashPosition?.quantity ?? 0;

    const addTransferPortfolio = cashAvailableInPortfolio + amount;
    const subtractTransfer = cashAvailable - amount;

    makeTransfer({
      variables: {
        investorId,
        portfolioId: portfolio.getId(),
        createFundTransferRequestInput: {
          quantity: amount,
          currency: "USD"
        }
      },
      update(cache) {
        cache.modify({
          id: `Position2:USD:${portfolio.getId()}`,
          fields: {
            settled() {
              return addTransferPortfolio;
            },
            quantity() {
              return addTransferPortfolio;
            }
          }
        });

        cache.modify({
          id: `Position2:USD:${cashPortfolio.portfolioId}`,
          fields: {
            settled() {
              return subtractTransfer;
            },
            quantity() {
              return subtractTransfer;
            }
          }
        });
      }
    });
  };
}

export default useDeposit;
