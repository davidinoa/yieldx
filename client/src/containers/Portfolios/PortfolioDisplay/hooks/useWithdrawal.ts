import {
  usePostInvestorPortfolioInternalPortfolioFundTransferMutation,
  Position2
} from "providers/graphql/hooks";

import YieldXPortfolio from "models/YieldXPortfolio";

interface Params {
  investorId: string;
  amount: number;
  cashPortfolio: {
    cashPos: Position2;
    cashAvailable: number;
    portfolioId: string;
  };
  portfolio: YieldXPortfolio;
}

function useWithdrawal() {
  const [
    makeTransfer
  ] = usePostInvestorPortfolioInternalPortfolioFundTransferMutation();

  return ({ investorId, amount, portfolio, cashPortfolio }: Params) => {
    const { cashAvailable } = cashPortfolio;
    const cashPosition = portfolio.positions.find(
      position => position.id === "CASH-USD" || position.id === "USD"
    );
    const cashAvailableInPortfolio =
      (cashPosition?.unsettled ?? 0) > 0
        ? cashPosition?.settled ?? 0
        : cashPosition?.quantity ?? 0;

    const addTransfer = cashAvailable + amount;
    const subtractTransferPortfolio = cashAvailableInPortfolio - amount;
    makeTransfer({
      variables: {
        investorId,
        portfolioId: portfolio.getId(),
        createFundTransferRequestInput: {
          quantity: amount * -1,
          currency: "USD"
        }
      },
      update(cache) {
        cache.modify({
          id: `Position2:USD:${portfolio.getId()}`,
          fields: {
            settled() {
              return subtractTransferPortfolio;
            },
            quantity() {
              return subtractTransferPortfolio;
            }
          }
        });

        cache.modify({
          id: `Position2:USD:${cashPortfolio.portfolioId}`,
          fields: {
            settled() {
              return addTransfer;
            },
            quantity() {
              return addTransfer;
            }
          }
        });
      }
    });
  };
}

export default useWithdrawal;
