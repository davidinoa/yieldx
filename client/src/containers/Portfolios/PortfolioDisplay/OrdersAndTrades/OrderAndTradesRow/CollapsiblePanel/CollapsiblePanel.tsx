import React from "react";
import dayjs from "dayjs";
import { Col, Row } from "react-bootstrap";
import { AssetAssetTypeEnum } from "@bondhouse/rover-universe";
import { displayDate } from "containers/Users/UserList/UserList";
import { currencyFormatter } from "containers/Investors/InvestorAccount/InvestorPortfolios/InvestorPortfolios";
import { LedgerEntry, Order2, OrderStatus } from "providers/graphql/hooks";
import { ApolloError } from "@apollo/client";
import { Spinner } from "elements/Spinner/Spinner";
import {
  CollapsibleRow,
  Heading,
  HeadingColumn,
  HeadRow,
  MoreInfoValue as Value,
  MoreInfoWrapper as Wrapper,
  StatusBadge,
  TradesHeader,
  TradesList,
  TradesWrapper
} from "containers/Portfolios/PortfolioDisplay/OrdersAndTrades/OrdersAndTrades.styles";

type Props = {
  order: Order2;
  trades: {
    data?: Array<LedgerEntry> | null;
    loading: boolean;
    error?: ApolloError;
  };
  assetType?: AssetAssetTypeEnum;
};

export default function CollapsiblePanel({ order, trades, assetType }: Props) {
  const isClosedSell =
    order.status === OrderStatus.Closed && order.quantity !== 0;
  const isCanceledBuy =
    order.status !== OrderStatus.Canceled && order.quantity !== 0;

  const hasTrades = !trades.loading && trades.data && trades.data.length !== 0;

  const hasNoTrades =
    !trades.loading && trades.data && trades.data.length === 0;
  return (
    <span data-testid={`OrderAndTradesRow-CollapsiblePanel-${order.id}`}>
      <CollapsibleRow>
        <Col className="order-info">
          <ul>
            <Heading>Order info:</Heading>
            <Wrapper>
              <div className="label">order type:</div>
              {order.fullFillRequired ? "All or none" : "Partial fill"}
            </Wrapper>
            <Wrapper>
              <div className="label">good until cancel:</div>
              {order.goodTillCancel ? "Yes" : "No"}
            </Wrapper>
            <Wrapper>
              <div className="label">created at:</div>
              <Value>{order.createdAt && displayDate(order.createdAt)}</Value>
            </Wrapper>
            <Wrapper>
              <div className="label">updated at:</div>
              <Value>{order.updatedAt && displayDate(order.updatedAt)}</Value>
            </Wrapper>
          </ul>
        </Col>
        <Col className="trades" xs={{ span: 16 }}>
          <TradesWrapper>
            <TradesHeader>
              <Heading>Trades:</Heading>
              <HeadRow>
                <HeadingColumn>status</HeadingColumn>
                <HeadingColumn>quantity</HeadingColumn>
                <HeadingColumn>buy/sell</HeadingColumn>
                <HeadingColumn>fill price</HeadingColumn>
                <HeadingColumn>total amount</HeadingColumn>
                <HeadingColumn>updated</HeadingColumn>
              </HeadRow>
            </TradesHeader>
            <TradesList>
              {trades.loading && <Spinner />}
              {(isClosedSell || isCanceledBuy) &&
                hasTrades &&
                trades.data?.map(trade => (
                  <Row key={trade.id}>
                    <Col data-testid="CollapsiblePanel-TradesList-Status">
                      <StatusBadge status={trade.status}>
                        {trade.status}
                      </StatusBadge>
                    </Col>
                    <Col>{Math.abs(trade.quantity)}</Col>
                    <Col data-testid="CollapsiblePanel-TradesList-Type">
                      {trade.quantity > 0 ? "BUY" : "SELL"}
                    </Col>
                    <Col>{currencyFormatter.format(trade.tradePrice)}</Col>
                    <Col>
                      {currencyFormatter.format(
                        Math.abs(
                          (assetType === AssetAssetTypeEnum.Bond ||
                          assetType?.toString() === "BOND"
                            ? trade.tradePrice / 100
                            : trade.tradePrice) * trade.quantity
                        )
                      )}
                    </Col>
                    <Col>
                      {dayjs(
                        trade.updatedAt ?? new Date().toISOString()
                      ).format("MMM D, YYYY")}
                    </Col>
                  </Row>
                ))}
              {(isClosedSell || isCanceledBuy) && hasNoTrades && (
                <p
                  style={{ fontSize: 13, height: "100%" }}
                  className="d-flex align-items-center justify-content-center"
                >
                  <span>No trades available</span>
                </p>
              )}
            </TradesList>
          </TradesWrapper>
        </Col>
      </CollapsibleRow>
    </span>
  );
}
