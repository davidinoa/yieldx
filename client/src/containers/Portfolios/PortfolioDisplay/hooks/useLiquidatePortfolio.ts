import {
  Direction2,
  Order2,
  OrderManagementGetOrderDocument,
  OrderStatus,
  PortfolioStatus,
  usePatchPortfolioMutation,
  usePostPortfolioOrdersMutation
} from "providers/graphql/hooks";

import YieldXPortfolio from "models/YieldXPortfolio";
import useCancelOrder from "containers/Portfolios/PortfolioDisplay/hooks/useCancelOrder";

function useLiquidatePortfolio() {
  const [updatePortfolio] = usePatchPortfolioMutation();
  const [postOrder] = usePostPortfolioOrdersMutation();
  const cancelOrder = useCancelOrder();

  return async (portfolio: YieldXPortfolio, orders: Array<Order2>) => {
    orders.forEach(order => {
      const { id: orderId, status } = order ?? {};
      if (!orderId) throw Error("No order ID");
      if (status === OrderStatus.Executing)
        throw Error("Cannot liquidate portfolio with executing orders");
      if (status === OrderStatus.Open) cancelOrder(orderId, portfolio.getId());
    });
    const referencePricesList =
      portfolio?.positions?.map(it => it.asset.lastPrice) ?? [];
    const referencePricesMap = referencePricesList.reduce<{
      [assetId: string]: number | null;
    }>((acc, cv, i) => {
      acc[portfolio?.positions[i]?.assetId ?? ""] = !cv ? null : cv;
      return acc;
    }, {});
    await Promise.all(
      portfolio.positions.map(position =>
        postOrder({
          variables: {
            portfolioId: portfolio.getId(),
            createOrderRequestInput: {
              assetId: position.assetId,
              quantity: position.quantity,
              goodTillCancel: true,
              direction: Direction2.Sell,
              priceLimit: (position.asset?.price ?? 1) * 0.98,
              spotPrice:
                referencePricesMap[position.assetId] ??
                position.asset.price ??
                undefined
            }
          },
          update(cache, { data }) {
            const newOrder = data?.postPortfolioOrders?.order;
            if (newOrder)
              cache.writeQuery({
                data: newOrder,
                query: OrderManagementGetOrderDocument,
                variables: {
                  id: newOrder.id,
                  portfolioId: newOrder.portfolioId
                }
              });
          }
        })
      )
    );

    await updatePortfolio({
      variables: {
        id: portfolio.getId(),
        updatePortfolioRequestInput: {
          status: PortfolioStatus.Terminated
        }
      },
      update(cache) {
        cache.modify({
          id: `Portfolio3:${portfolio.getId()}`,
          fields: {
            status() {
              return PortfolioStatus.Terminated;
            }
          }
        });
      }
    });
  };
}

export default useLiquidatePortfolio;
