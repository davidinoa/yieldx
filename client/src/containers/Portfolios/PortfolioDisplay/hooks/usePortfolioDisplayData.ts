import { useHistory, useLocation, useParams } from "react-router-dom";
import {
  Order2,
  PortfolioType,
  Position2,
  useOrderManagementGetOrdersQuery
} from "providers/graphql/hooks";
import useGetYieldxPortfolio from "providers/graphql/hooks/useGetYieldxPortfolio";
import useGetPortfolioInvestor from "providers/graphql/hooks/useGetPortfolioInvestor";
import useGetInvestorCashPortfolio from "providers/graphql/hooks/useGetInvestorCashPortfolio";
import useCancelOrder from "containers/Portfolios/PortfolioDisplay/hooks/useCancelOrder";
import useDeleteExternalPortfolio from "providers/graphql/hooks/useDeleteExternalPortfolio";
import useLiquidatePortfolio from "containers/Portfolios/PortfolioDisplay/hooks/useLiquidatePortfolio";
import useDeposit from "containers/Portfolios/PortfolioDisplay/hooks/useDeposit";
import useWithdrawal from "containers/Portfolios/PortfolioDisplay/hooks/useWithdrawal";
import {
  AlertLevel,
  AlertMessage,
  useAlertContext
} from "providers/alerts/AlertProvider";

type PortfolioDisplayParams = {
  portfolioId: string;
  type: string;
};

export default function usePortfolioDisplayData() {
  const { type, portfolioId } = useParams<PortfolioDisplayParams>();
  const portfolioType =
    type === "external" ? PortfolioType.External : PortfolioType.Live;

  const history = useHistory();
  const { pathname, hash } = useLocation();

  const {
    portfolio,
    loading: portfolioLoading,
    error: portfolioError
  } = useGetYieldxPortfolio({ portfolioId, portfolioType });
  if (portfolioError) console.error(portfolioError);
  const {
    data: ordersData,
    loading: ordersLoading,
    error: ordersError
  } = useOrderManagementGetOrdersQuery({
    variables: {
      portfolioId
    },
    skip: portfolioType === "EXTERNAL" || !portfolio
  });
  const orders = ordersData?.orderManagementGetOrders?.orders;
  const cancelOrder = useCancelOrder();
  const deleteExternal = useDeleteExternalPortfolio();
  const {
    investor,
    error: investorError,
    loading: investorLoading
  } = useGetPortfolioInvestor(
    portfolioId,
    !portfolio || portfolioType !== PortfolioType.Live
  );

  const {
    data: { cashPortfolio },
    error: cashPortfolioError,
    loading: cashPortfolioLoading
  } = useGetInvestorCashPortfolio(investor?.id);

  const liquidatePortfolio = useLiquidatePortfolio();
  const makeDeposit = useDeposit();
  const makeWithdrawal = useWithdrawal();
  const { triggerAlert } = useAlertContext();
  return {
    portfolio,
    portfolioLoading,
    portfolioError,
    orders,
    ordersLoading,
    ordersError,
    investorError,
    investorLoading,
    cashPortfolioError,
    cashPortfolioLoading,
    history,
    pathname,
    hash,
    investorCashAvailable: cashPortfolio?.cashAvailable,
    onDelete: async () => {
      try {
        await deleteExternal(portfolioId).then(() => {
          triggerAlert(
            "You portfolio has been closed successfully",
            AlertLevel.success,
            { timeout: 2500 }
          );
        });
      } catch (err) {
        console.error(err);
        triggerAlert(AlertMessage.Unexpected, AlertLevel.danger);
      }
      history.push("/portfolios#alert=PortfolioDeleted");
    },
    onCancelOrder: async (orderId: string) => {
      try {
        await cancelOrder(orderId, portfolioId);
        history.push(`${pathname}#orders`);
      } catch (e) {
        console.error(e);
      }
    },
    onLiquidate: async () => {
      const investorId = investor?.id;
      if (!investorId) throw Error("Investor ID is missing");
      if (!portfolio)
        throw Error("Portfolio is undefined. Failed to liquidate");
      if (!orders) throw Error("Orders undefined, failed to liquidate");
      await liquidatePortfolio(portfolio, orders as Array<Order2>);
      history.push(`/investors/${investorId}#portfolios`, {
        shouldRefresh: true
      });
    },

    onDeposit: async (amount: number) => {
      const investorId = investor?.id;
      if (!investorId) throw Error("Investor ID is missing");
      if (!portfolio) throw Error("Portfolio missing");
      if (!cashPortfolio) throw Error("Cash investor/portfolio is missing");
      makeDeposit({
        investorId,
        portfolio,
        amount,
        cashPortfolio: cashPortfolio as {
          cashPos: Position2;
          cashAvailable: number;
          portfolioId: string;
        }
      });
      history.push(pathname);
    },

    onWithdraw: async (amount: number) => {
      const investorId = investor?.id;
      if (!investorId) throw Error("Investor ID is missing");
      if (!portfolio) throw Error("Portfolio missing");
      if (!cashPortfolio) throw Error("Cash investor/portfolio is missing");
      makeWithdrawal({
        investorId,
        portfolio,
        amount,
        cashPortfolio: cashPortfolio as {
          cashPos: Position2;
          cashAvailable: number;
          portfolioId: string;
        }
      });
      history.push(pathname);
    }
  };
}
