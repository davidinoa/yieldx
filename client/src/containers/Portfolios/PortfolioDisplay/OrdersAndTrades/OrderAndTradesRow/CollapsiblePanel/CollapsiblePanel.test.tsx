import React from "react";
import { render, screen } from "test/utils";
import { Order2, OrderStatus, Status3 } from "providers/graphql/hooks";
import { AssetAssetTypeEnum } from "@bondhouse/rover-universe";
import CollapsiblePanel from "./CollapsiblePanel";

const mockOrder: Order2 = {
  id: "5d62a607-9c33-4610-9310-eeb45351c37a",
  portfolioId: "6cdf10c3-000c-40c5-886e-c45f4b8226f7",
  quantity: 1,
  assetId: "4207ed32-9c48-4235-9821-f0612f54a009",
  createdAt: "2020-07-15T15:00:02.250811Z",
  updatedAt: "2020-07-15T15:00:02.250811Z",
  status: OrderStatus.Open
};

const mockTrades = {
  data: [
    {
      assetId: "4a81cf99-eca3-4d99-a922-a0ac88e1bcb9",
      comment: null,
      counterPartyId: "TRAFIX",
      createdAt: "2020-10-09T14:19:57.005809Z",
      fee: null,
      id: "63daac2b-f677-4a29-a95f-312470567261",
      metadata: null,
      orderId: "357d5ac2-79bd-4c03-8cdd-342f943e3575",
      portfolioId: "9b68a035-910d-41d8-8452-2f7175f13ae2",
      quantity: 200,
      relatedPositions: [
        {
          positionId: "7675d7d4-2799-4df1-82ec-fd9bd071a76f",
          quantity: 200
        }
      ],
      status: Status3.Confirmed,
      tradeId: "df2bae71-3537-407a-94bb-6cc3b8786f0e",
      tradePrice: 16.09,
      updatedAt: "2020-10-09T14:19:57.005809Z"
    }
  ],
  loading: false
};

describe("CollapsiblePanel", () => {
  test("Should correctly render no trades", () => {
    const { asFragment } = render(
      <CollapsiblePanel
        order={mockOrder}
        trades={{
          data: [],
          loading: false
        }}
        assetType={"FUND" as AssetAssetTypeEnum.Fund}
      />
    );
    expect(screen.getByText(/order info/i)).toBeInTheDocument();
    // Should render no trades
    expect(
      screen.queryAllByTestId("CollapsiblePanel-TradesList-Status").length
    ).toBe(0);
    expect(asFragment()).toMatchSnapshot();
  });

  test("Should render 1 trade", () => {
    const { asFragment } = render(
      <CollapsiblePanel
        order={mockOrder}
        trades={mockTrades}
        assetType={"FUND" as AssetAssetTypeEnum.Fund}
      />
    );
    expect(
      screen.getAllByTestId("CollapsiblePanel-TradesList-Status").length
    ).toBe(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
