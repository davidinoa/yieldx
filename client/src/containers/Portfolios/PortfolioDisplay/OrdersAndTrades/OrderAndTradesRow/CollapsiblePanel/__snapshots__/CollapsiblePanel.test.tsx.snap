// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsiblePanel Should correctly render no trades 1`] = `
<DocumentFragment>
  .c0 {
  background: var(--dark);
  margin-top: -7px;
  margin-bottom: 30px;
}

.c0 ul {
  padding: 0;
}

.c0 .order-info {
  padding: 4.5rem 5rem;
  padding-right: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 1px solid #253859;
}

.c0 .order-info h4 {
  margin-bottom: 1.5rem;
}

.c0 .order-info li {
  font-size: 14px;
}

.c0 .order-info li .label {
  padding-right: 0.5rem;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 11px;
  line-height: 13px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
  color: #6980a6;
}

.c0 .trades {
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  margin-bottom: 15px;
  display: grid;
  grid-template-columns: 2fr 3fr;
}

.c3 {
  text-transform: capitalize;
  font-size: 14px;
  line-height: 17px;
  color: #ffffff;
}

.c5 {
  width: 100%;
  background: #192b4a;
  border: 1px solid #253859;
  padding: 1rem;
}

.c1 {
  font-weight: bold;
  font-size: 15px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 {
  font-weight: bold;
  font-size: 11px;
  line-height: 13px;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
  color: var(--lightest);
  text-transform: uppercase;
  text-align: center;
}

.c8 {
  border: 1px solid #253859;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: top;
  -webkit-justify-content: top;
  -ms-flex-pack: top;
  justify-content: top;
  padding: 1rem;
  background: var(--dark);
  height: 100%;
  width: 100%;
  text-align: center;
  font-size: 13px;
}

.c4 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4:last-child {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c4 > * {
  width: 100%;
}

<span
    data-testid="OrderAndTradesRow-CollapsiblePanel-5d62a607-9c33-4610-9310-eeb45351c37a"
  >
    <div
      class="c0 row"
    >
      <div
        class="order-info col"
      >
        <ul>
          <h4
            class="c1"
          >
            Order info:
          </h4>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              order type:
            </div>
            Partial fill
          </li>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              good until cancel:
            </div>
            No
          </li>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              created at:
            </div>
            <span
              class="c3"
            >
              July 15, 2020
            </span>
          </li>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              updated at:
            </div>
            <span
              class="c3"
            >
              July 15, 2020
            </span>
          </li>
        </ul>
      </div>
      <div
        class="trades col-16"
      >
        <div
          class="c4"
        >
          <header
            class="c5"
          >
            <h4
              class="c1"
            >
              Trades:
            </h4>
            <div
              class="c6"
            >
              <div
                class="c7 col"
              >
                status
              </div>
              <div
                class="c7 col"
              >
                quantity
              </div>
              <div
                class="c7 col"
              >
                buy/sell
              </div>
              <div
                class="c7 col"
              >
                fill price
              </div>
              <div
                class="c7 col"
              >
                total amount
              </div>
              <div
                class="c7 col"
              >
                updated
              </div>
            </div>
          </header>
          <div
            class="c8"
          >
            <p
              class="d-flex align-items-center justify-content-center"
              style="font-size: 13px; height: 100%;"
            >
              <span>
                No trades available
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CollapsiblePanel Should render 1 trade 1`] = `
<DocumentFragment>
  .c0 {
  background: var(--dark);
  margin-top: -7px;
  margin-bottom: 30px;
}

.c0 ul {
  padding: 0;
}

.c0 .order-info {
  padding: 4.5rem 5rem;
  padding-right: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 1px solid #253859;
}

.c0 .order-info h4 {
  margin-bottom: 1.5rem;
}

.c0 .order-info li {
  font-size: 14px;
}

.c0 .order-info li .label {
  padding-right: 0.5rem;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 11px;
  line-height: 13px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
  color: #6980a6;
}

.c0 .trades {
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c9 {
  --color-open: #fae369;
  --color-pending: #fae369;
  --color-executing: #0fd4ff;
  --color-confirmed: #42f1d2;
  --color-closed: var(--white);
  --color-settled: var(--white);
  --color-canceled: #f87584;
  --background-open: #fae36910;
  --background-pending: #fae36910;
  --background-executing: #0fd4ff10;
  --background-confirmed: #42f1d210;
  --background-closed: #ffffff10;
  --background-settled: #ffffff10;
  --background-canceled: #f8758410;
  color: var(--color-confirmed);
  background: var(--background-confirmed);
  padding: 5px 25px;
  border-radius: 50px;
  text-align: center;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 10px;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  width: 100%;
  line-height: initial;
}

.c2 {
  margin-bottom: 15px;
  display: grid;
  grid-template-columns: 2fr 3fr;
}

.c3 {
  text-transform: capitalize;
  font-size: 14px;
  line-height: 17px;
  color: #ffffff;
}

.c5 {
  width: 100%;
  background: #192b4a;
  border: 1px solid #253859;
  padding: 1rem;
}

.c1 {
  font-weight: bold;
  font-size: 15px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 {
  font-weight: bold;
  font-size: 11px;
  line-height: 13px;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
  color: var(--lightest);
  text-transform: uppercase;
  text-align: center;
}

.c8 {
  border: 1px solid #253859;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: top;
  -webkit-justify-content: top;
  -ms-flex-pack: top;
  justify-content: top;
  padding: 1rem;
  background: var(--dark);
  height: 100%;
  width: 100%;
  text-align: center;
  font-size: 13px;
}

.c4 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4:last-child {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c4 > * {
  width: 100%;
}

<span
    data-testid="OrderAndTradesRow-CollapsiblePanel-5d62a607-9c33-4610-9310-eeb45351c37a"
  >
    <div
      class="c0 row"
    >
      <div
        class="order-info col"
      >
        <ul>
          <h4
            class="c1"
          >
            Order info:
          </h4>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              order type:
            </div>
            Partial fill
          </li>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              good until cancel:
            </div>
            No
          </li>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              created at:
            </div>
            <span
              class="c3"
            >
              July 15, 2020
            </span>
          </li>
          <li
            class="c2"
          >
            <div
              class="label"
            >
              updated at:
            </div>
            <span
              class="c3"
            >
              July 15, 2020
            </span>
          </li>
        </ul>
      </div>
      <div
        class="trades col-16"
      >
        <div
          class="c4"
        >
          <header
            class="c5"
          >
            <h4
              class="c1"
            >
              Trades:
            </h4>
            <div
              class="c6"
            >
              <div
                class="c7 col"
              >
                status
              </div>
              <div
                class="c7 col"
              >
                quantity
              </div>
              <div
                class="c7 col"
              >
                buy/sell
              </div>
              <div
                class="c7 col"
              >
                fill price
              </div>
              <div
                class="c7 col"
              >
                total amount
              </div>
              <div
                class="c7 col"
              >
                updated
              </div>
            </div>
          </header>
          <div
            class="c8"
          >
            <div
              class="row"
            >
              <div
                class="col"
                data-testid="CollapsiblePanel-TradesList-Status"
              >
                <div
                  class="c9"
                >
                  CONFIRMED
                </div>
              </div>
              <div
                class="col"
              >
                200
              </div>
              <div
                class="col"
                data-testid="CollapsiblePanel-TradesList-Type"
              >
                BUY
              </div>
              <div
                class="col"
              >
                $16.09
              </div>
              <div
                class="col"
              >
                $3,218.00
              </div>
              <div
                class="col"
              >
                Oct 9, 2020
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </span>
</DocumentFragment>
`;
